sdk/azcore/runtime/policy_bearer_token.go

```diff
@@ -115,9 +124,80 @@ func (b *BearerTokenPolicy) Do(req *policy.Request) (*http.Response, error) {
 	return res, err
 }
 
+func (b *BearerTokenPolicy) handleClaimsChallenge(_ *policy.Request, res *http.Response, authNZ func(policy.TokenRequestOptions) error) error {
+	claims, err := parseCAEChallenge(res)
+	if err != nil {
+		// challenge contains claims we can't parse
+		return err
+	}
+	if claims == "" {
+		// no claims in challenge, so this is a simple authorization failure
+		return NewResponseError(res)
+	}
+	// request a new token having the specified claims, send the request again
+	return authNZ(policy.TokenRequestOptions{Claims: claims, EnableCAE: true, Scopes: b.scopes})
+}
+
 func checkHTTPSForAuth(req *policy.Request, allowHTTP bool) error {
 	if strings.ToLower(req.Raw().URL.Scheme) != "https" && !allowHTTP {
 		return errorinfo.NonRetriableError(errors.New("authenticated requests are not permitted for non TLS protected (https) endpoints"))
 	}
 	return nil
 }
+
+var (
+	claimsChallenge *regexp.Regexp
+	once            = &sync.Once{}
+)
+
+// parseCAEChallenge extracts claims from the first CAE challenge in the response's WWW-Authenticate header. If
+// the response doesn't contain a CAE challenge, it returns an empty string and nil error. It returns a non-nil
+// error only when it identifies a CAE challenge whose claims aren't valid base64.
+func parseCAEChallenge(res *http.Response) (string, error) {
+	once.Do(func() {
+		// This expression matches CAE claims challenges and captures their parameters. It doesn't
+		// correspond precisely to the challenge grammar in RFC 7235 appendix C because CAE challenges
+		// are more narrowly defined at
+		// https://learn.microsoft.com/entra/identity-platform/claims-challenge#claims-challenge-header-format
+		claimsChallenge = regexp.MustCompile(`(?:\w+ ((?:\w+="[^"]*",?\s*)+))`)
+	})
+	// WWW-Authenticate can have multiple values, each containing multiple challenges
+	for _, h := range res.Header.Values(shared.HeaderWWWAuthenticate) {
+		for _, sm := range claimsChallenge.FindAllStringSubmatch(h, -1) {
+			if len(sm) < 2 {
+				continue
+			}
+			for _, params := range sm[1:] {
+				cae := false
+				claims := ""
+				for _, param := range strings.Split(params, ", ") {
+					k, v, found := strings.Cut(param, `="`)
+					if !found {
+						// should never happen given the regex that got us here
+						continue
+					}
+					switch strings.Trim(k, " ") {
+					case "claims":
+						claims = v
+					case "error":
+						// v has an irrelevant trailing quote
+						if strings.HasPrefix(v, "insufficient_claims") {
+							cae = true
+						}
+					}
+				}
+				if cae && claims != "" {
+					// remove the trailing quote and any base64 padding
+					if end := strings.IndexAny(claims, `="`); end > 0 {
+						claims = claims[:end]
+					}
+					if b, de := base64.RawURLEncoding.DecodeString(claims); de == nil {
```

Comment:
Are we expected to decode the claims rather than just pass them along as is?

Reply:
That depends on the contract with MSAL. MSAL Go expects the decoded claims i.e. JSON.